DataSetExplorer/Core/DataSets/CaDETToDataSetProjectBuilder.cs

```diff
@@ -110,9 +111,85 @@ private int DetermineNumberOfInstances(int totalNumber)
 
         private List<Instance> CaDETToDataSetProjectClasses(List<CaDETClass> cadetClasses)
         {
+            CreateReferencesMap(cadetClasses);
             return cadetClasses.Select(c => new Instance(
-                c.FullName, GetCodeUrl(c.FullName), _projectAndCommitUrl, SnippetType.Class, _cadetProject.GetMetricsForCodeSnippet(c.FullName)
-            )).ToList();
+                    c.FullName, GetCodeUrl(c.FullName), _projectAndCommitUrl, SnippetType.Class,
+                    _cadetProject.GetMetricsForCodeSnippet(c.FullName), FindRelatedInstances(c)
+                )).ToList();
+        }
+
+        private void CreateReferencesMap(List<CaDETClass> cadetClasses)
+        {
+            foreach (var instance in cadetClasses)
+            {   
+                foreach (var reference in GetReferencedInstances(instance))
+                {
+                    AddReferenceToMap(reference, instance);
+                }
+            }
+        }
+
+        private static Dictionary<CaDETClass, int> GetReferencedInstances(CaDETClass c)
+        {
+            var referencedInstances = new List<CaDETClass>();
+            referencedInstances.AddRange(c.GetFieldLinkedTypes());
```

Comment:
My issue with understanding this code is that you are using "reference" to mean both directions of access.

My algorithm for finding a good name includes brainstorming a description of the two concepts and than seeing which keywords to use for the final name.

On the one hand, we have "classes that our class references" (i.e., the referencedInstances in this code), and on the other we have the "classes that reference our class".

Maybe we can separate the terms to have "ReferencedInstances(CaDETClass referencingClass)" and "InstancesThatReference(CaDETClass referencedClass)". Revisit the terminology in your functions to see if every method talks about the correct "reference".